### PR TITLE
Fix a range of new byte compiler warnings

### PR DESCRIPTION
**Pull request by jabranham**
Emacs built from git master adds a few more byte compilation checks,
this addresses them in preparation for the ESS 18.09 release:

- Aliases should be defined _before_ their referent. I moved them up,
  except for those that have been obsolete for five years or more. Those got
  removed. (there were only two or three of these)

- The byte compiler now checks whether cl-lib is needed at runtime,
  not just compile time so added where needed

- set-buffer-process-coding-system is now marked for interactive use
  only, so prefer set-process-coding-system

### CODE REVIEW

File: lisp/ess-trns.el [hunk 1]
```diff
@@ -162,7 +162,7 @@ in the region, leaving only the S commands.  Other keybindings are:
   (setq comint-use-prompt-regexp t)
   (make-local-variable 'comint-prompt-regexp)
   (setq comint-prompt-regexp (concat "^" inferior-ess-prompt))
-  (setq inferior-font-lock-keywords 'inferior-R-font-lock-keywords)
+  (setq inferior-ess-font-lock-keywords 'inferior-R-font-lock-keywords)
```
Comment (jabranham) [inline lisp/ess-trns.el]: @vspinu  - why did you change this recently? Won't that mess up font lock for all non-R inferior modes started from transcript buffers?

Comment (vspinu) [reply lisp/ess-trns.el]: I didn't change anything substantially here (except of introducing that typo which you fixed). See f51df5dcfc9bf Transcript is in R output only and `inferior-ess-font-lock` is buffer local. Or do you have other issues in mind here?

**Comment by vspinu** [comment]
> Emacs built from git master adds a few more byte compilation checks,

Some of those might not stay till release. The referal one is a special pain in the ass because it affect functions only but not variables or other obsoletes.

In any case, sharp eye and good job! Thanks.

File: lisp/ess-trns.el (hunk 1 above)
Comment (vspinu) [reply lisp/ess-trns.el]: I don't know, but that was the old behavior and I am reasonably sure almost no one is using transcript even for R, let alone stata and sas.
